src/sessions/session_manual.cpp

```diff
@@ -189,15 +189,30 @@ void session_manual::attach_protocols(channel::ptr channel)
 }
 
 void session_manual::handle_channel_stop(const code& ec,
-    const std::string& hostname, uint16_t port)
+    const std::string& hostname, uint16_t port, uint32_t remaining,
+    channel_handler handler)
 {
     LOG_DEBUG(LOG_NETWORK)
         << "Manual channel stopped: " << ec.message();
 
     // Special case for already connected, do not keep trying.
     // After a stop we don't use the caller's start handler, but keep connecting.
     if (ec != error::address_in_use)
-        connect(hostname, port);
+    {
+        // Retry forever if limit is zero.
+        remaining = settings_.manual_attempt_limit == 0 ? 1 : remaining;
+
+        if (remaining > 0)
+        {
+            start_connect(error::success, hostname, port, remaining, handler);
+            return;
+        }
+
+        LOG_WARNING(LOG_NETWORK)
+            << "Not restarting manual connection to ["
+            << config::endpoint(hostname, port) << "] after "
+            << settings_.manual_attempt_limit << " failed attempts.";
```

Comment:
I think the stop handler must either forward or invoke `handler`, as the guarantee is that it is always fired once, either on successful start or on failure to start (which should be a stop error). This applies to all returns from stop.

Reply:
Actually I think the right way to handle this is to invoke `handler` only and always in the start handler, leaving the stop handler to loop implantation.

Reply:
Must've dropped it while I was retrying some variations and started getting actual connection failures.

Reply:
See my last comment.

Reply:
Extended handle_channel_start with remaining and conditionally invoke handler on error in both not retrying and `ec == error::address_in_use`.

Reply:
Handler must also be invoked with success.

Reply:
Still is.

Reply:
Sorry, just a consequence of reviewing on my phone, missed it.